src/Limiters/limiter.jl

```diff
@@ -0,0 +1,119 @@
+
+    # Compute ρ's and ρq's masses over an element e
+    for e in 1:n_elems
+        ρ_e_mass = sum(Fields.slab(ρ, e))
+        ρq_e_mass = sum(Fields.slab(ρq, e))
+
+        # This should never happen, but if it does, don't limit
+        if ρ_e_mass <= 0
+            continue
+        end
+
+        # Relax constraints to ensure limiter has a solution:
+        # This is only needed if runnign with the SSP CFL>1 or
+        # due to roundoff errors.
+        if ρq_e_mass < min_ρq[e] * ρ_e_mass
+            min_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        if ρq_e_mass > max_ρq[e] * ρ_e_mass
+            max_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        local_geometry_slab = Fields.slab(space.local_geometry, e)
+
+        q = parent(Fields.slab(ρq, e)) ./ parent(Fields.slab(ρ, e))
+        for iter in 1:(Nq * Nq - 1)
+            mass_change = 0.0
+            # Iterate over quadrature points
+            for j in 1:Nq, i in 1:Nq
+                ρ_wJ[i, j] =
+                    local_geometry_slab[i, j].WJ * parent(ρ)[i, j, 1, e]
+                # Compute the error tolerance and project q into the
+                # upper and lower bounds
+                if q[i, j, 1] > max_ρq[e]
+                    mass_change += (q[i, j, 1] - max_ρq[e]) * ρ_wJ[i, j]
+                    q[i, j, 1] = max_ρq[e]
+                elseif q[i, j, 1] < min_ρq[e]
+                    mass_change -= (min_ρq[e] - q[i, j, 1]) * ρ_wJ[i, j]
+                    q[i, j, 1] = min_ρq[e]
+                end
+            end
+
+            # By this projection of q to the upper/lower bounds, we
+            # have introduced a mass_change. Check if we are within the chosen
+            # tolerance
+            if abs(mass_change) <= rtol * abs(ρq_e_mass)
+                break
+            end
+
+            weights_sum = 0.0
+
+            # If the change was positive, the removed mass is added
+            if mass_change > 0
+                # Iterate over quadrature points
+                for j in 1:Nq, i in 1:Nq
+                    if q[i, j, 1] < max_ρq[e]
+                        weights_sum += ρ_wJ[i, j]
+                    end
+                end
+                for j in 1:Nq, i in 1:Nq
+                    if q[i, j, 1] < max_ρq[e]
+                        q[i, j, 1] += mass_change / weights_sum
+                    end
+                end
+            else # If the change was negative, the added mass is removed
+                # Iterate over quadrature points
+                for j in 1:Nq, i in 1:Nq
+                    if q[i, j, 1] > min_ρq[e]
+                        weights_sum += ρ_wJ[i, j]
+                    end
+                end
+                for j in 1:Nq, i in 1:Nq
+                    if q[i, j, 1] > min_ρq[e]
+                        q[i, j, 1] += mass_change / weights_sum
+                    end
+                end
+            end
+        end
+        parent(Fields.slab(ρq, e)) .= q .* parent(Fields.slab(ρ, e))
```

Comment:
@jakebolewski I tried all the suggested changes but when it gets here there is another Method error. See the simple example below:
```
julia> for e in 1:n_elems
           ρ_e_slab = Fields.slab(ρ, e)
           Q_e_slab = Fields.slab(Q, e)
           
           q = Q_e_slab ./ ρ_e_slab
           @show q
           @show q .* ρ_e_slab
       end
q = Float64-valued Field:
  [-0.2, 1.1, 1.0e-5, 1.0]
ERROR: MethodError: no method matching parent_array_type(::Type{StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}})
Closest candidates are:
  parent_array_type(::Type{<:Array{T}}) where T at ~/CliMA/ClimaCore.jl/src/DataLayouts/struct.jl:89
  parent_array_type(::Type{<:SubArray{T, N, A}}) where {T, N, A} at ~/CliMA/ClimaCore.jl/src/DataLayouts/struct.jl:90
  parent_array_type(::Type{<:CUDA.CuArray{T, N, B} where N}) where {T, B} at ~/CliMA/ClimaCore.jl/src/DataLayouts/cuda.jl:22
Stacktrace:
 [1] ClimaCore.DataLayouts.DataStyle(#unused#::Type{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}})
   @ ClimaCore.DataLayouts ~/CliMA/ClimaCore.jl/src/DataLayouts/broadcast.jl:25
 [2] Base.Broadcast.BroadcastStyle(#unused#::Type{ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}})
   @ ClimaCore.Fields ~/CliMA/ClimaCore.jl/src/Fields/broadcast.jl:17
 [3] combine_styles(c::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}})
   @ Base.Broadcast ./broadcast.jl:435
 [4] combine_styles(c1::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, c2::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}})
   @ Base.Broadcast ./broadcast.jl:436
 [5] broadcasted(::Function, ::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, ::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}})
   @ Base.Broadcast ./broadcast.jl:1303
 [6] macro expansion
   @ ./show.jl:1047 [inlined]
 [7] top-level scope
   @ ./REPL[36]:7
 [8] top-level scope
   @ ~/.julia/packages/CUDA/nYggH/src/initialization.jl:52
```

The creation of ` q = Q_e_slab ./ ρ_e_slab` as a point-wise division of two slab Fields is Ok, but then when I multiply it again for the other slab Field there is a Method Error. I am very confused at this point. It seems like to work around this I have to use parent indeed.

Reply:
I think you are just uncovering some bugs (looks like the hack to get dual numbers here broke something)

Reply:
let's fix the bugs :)

Reply:
https://github.com/CliMA/ClimaCore.jl/pull/484/files

Reply:
Thanks for the quick fix!

Reply:
@jakebolewski it seems like this was fixed for the `slab(data::IJFH{S, Nij}, h::Integer)` but not for the `slab(field::Field, inds...)`, which is the one I was using in this case

Reply:
```
# Initialize fields
ρ = ones(space)
q = ones(space)
ρq = ρ .* q

n_elems = Topologies.nlocalelems(space.topology)

for e in 1:n_elems
    ρ_e_slab = Fields.slab(ρ, e)
    ρq_e_slab = Fields.slab(ρq, e)
    q = ρq_e_slab ./ ρ_e_slab
    ρq_e_slab .= q .* ρ_e_slab
end

ERROR: ArgumentError: Broadcasted{Unknown} wrappers do not have a style assigned
Stacktrace:
 [1] Base.Broadcast.BroadcastStyle(#unused#::Type{Base.Broadcast.Broadcasted{Base.Broadcast.Unknown, Nothing, typeof(*), Tuple{ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}}}})
   @ Base.Broadcast ./broadcast.jl:231
 [2] combine_styles(c::Base.Broadcast.Broadcasted{Base.Broadcast.Unknown, Nothing, typeof(*), Tuple{ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}}})
   @ Base.Broadcast ./broadcast.jl:435
 [3] combine_styles(c1::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, c2::Base.Broadcast.Broadcasted{Base.Broadcast.Unknown, Nothing, typeof(*), Tuple{ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}}})
   @ Base.Broadcast ./broadcast.jl:436
 [4] materialize!(dest::ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, bc::Base.Broadcast.Broadcasted{Base.Broadcast.Unknown, Nothing, typeof(*), Tuple{ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, StaticArrays.MArray{Tuple{2, 2, 1}, Float64, 3, 4}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}, ClimaCore.Fields.Field{ClimaCore.DataLayouts.IJF{Float64, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}, ClimaCore.Spaces.SpectralElementSpaceSlab{ClimaCore.Spaces.Quadratures.GLL{2}, ClimaCore.DataLayouts.IJF{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.XYPoint{Float64}, Float64, StaticArrays.SMatrix{2, 2, Float64, 4}}, 2, SubArray{Float64, 3, Array{Float64, 4}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Int64}, true}}}}}})
   @ Base.Broadcast ./broadcast.jl:868
 [5] top-level scope
   @ ./REPL[17]:6
 [6] top-level scope
   @ ~/.julia/packages/CUDA/nYggH/src/initialization.jl:52

Reply:
I thought you were going to use the underlying datalayout because you need to index into it, but we were missing a method for the Field case: https://github.com/CliMA/ClimaCore.jl/pull/486

Reply:
It was two different use cases. This algorithm needed two level of granularity. This one was needed to address your [comment](https://github.com/CliMA/ClimaCore.jl/pull/426/files#r793766535) above. Thanks for fixing this one, too. 
Now the only thing that remains to make this a bit more readable is the ability to access indexing of Fields instead of Field data, so that we could avoid the `Fields.todata` call. But this could be updated here in a later PR.